### PR TITLE
Virtual multi-z

### PR DESCRIPTION
**Pull request by PowerfulBacon**
**TODO**
- [x] When 2 z-levels above a mob, if the mob stands still for a period of time they will just disappear and stop drawing entirely.
- [x] VV support for above and below.
- [x] Things are just randomly disappearing, why!!!

## About The Pull Request

You may now set any turf to be above/below any other turf and it will be handled correctly. This allows for virtual z-levels to be used as mulit-zs.

**Technical Changes:**

- Since it is significantly more difficult to calculate the maximum depth at a certain point, instead we just always start building the bottom up from the maximum depth of zcopy. This leaves empty planes in the middle but that is fine and can be handled properly.
image

- The z_depth variable now represents how far up in the stack the turf is.
- GET_TURF_ABOVE and GET_TURF_BELOW have been modified so that below and above will cached and can be modified to a different turf now.
- Removed above() and below() procs since they are duplicating existing behaviours.
- Added the link_region proc to link regions together.

## Why It's Good For The Game

This allows for significantly more advanced multi-z behaviours involving virtual-z levels. For example, you could now have a 10x10 basement in the middle of the station, shuttles could span multiple z-levels if a little bit more work was done to them.
I plan to use this in my mining rework to create sub-levels to asteroids. This means that the asteroids would be able to have basement levels that you can descend to even though the station is only a single-z.

image

## Testing Photographs and Procedure

image

image

image

(Lighting issues were present with previous zcopy as far as I am aware)

image

image

For whatever reason, if the lighting below the turf is blue then lighting on the upper floors will be a deep blue colour despite not having an open turf at that location. This behaviour is extremely strange and I was to document it since I believe this existed prior and isn't a regression.

**Now for the good stuff**

https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/3c8ca6d9-b4bd-4405-a9ea-a8683da297dc

_This is only a demo right now and I am still working on solving some minor issues such as disappearing items and people._

image

## Changelog
:cl:
code: Implements support for virtual multi-z levels.
/:cl:

### CODE REVIEW

File: code/__DEFINES/multi_z.dm [hunk 1]
```diff
@@ -9,7 +9,30 @@
...
+/// Attempt to get the turf above the provided one according to either
+/// the map config defined z-traits, or the assigned above turf
+#define GET_TURF_ABOVE(turf) \
+( \
+	!(turf) ? null : ( \
+		(turf).above || ( \
+			(turf).set_above(MAPPING_TURF_ABOVE(turf)) \
+		) \
+	) \
+)
```
Comment (PowerfulBacon) [inline code/__DEFINES/multi_z.dm]: TODO: We need a flag on turfs in order to identify when there is no turf below so that we don't keep going into set_above calls over and over due to not being able to cache a below value.

File: code/_compile_options.dm [hunk 2]
```diff
@@ -84,6 +84,8 @@
 #define FORCE_MAP "runtimestation"
 #endif
 
+#define FORCE_MAP "multiz_debug"
```
Comment (PowerfulBacon) [inline code/_compile_options.dm]: TODO: Remove this

File: code/controllers/subsystem/zcopy.dm [hunk 3]
```diff
@@ -343,7 +297,7 @@ SUBSYSTEM_DEF(zcopy)
 
 			var/override_depth
 			var/original_type = object.type
-			var/original_z = object.z
+			var/original_depth = GET_TURF_DEPTH(T)
```
Comment (PowerfulBacon) [inline code/controllers/subsystem/zcopy.dm]: This is the reason things are buggy and disappearing, we need the turf depth of T.below, not T

**Comment by PowerfulBacon** [comment]
Solved the disappearing people issue it seems.
image

**Comment by PowerfulBacon** [comment]
image

This looks wrong, need to pull some screenshots of how they are meant to look

**Comment by PowerfulBacon** [comment]
From what it seems like, things are randomly disappearing when they are moving from 1 z level to another that isn't directly positioned above/below the other.

File: code/modules/atmospherics/environmental/LINDA_system.dm [hunk 4]
```diff
@@ -49,18 +49,9 @@
...
-			(direction & UP) ? \
-				(z_traits[Z_LEVEL_UP]) ? \
-					(get_step(locate(x, y, z + 1), NONE)) : \
-				(null) : \
-				(z_traits[Z_LEVEL_DOWN]) ? \
-					(get_step(locate(x, y, z - 1), NONE)) : \
-				(null) : \
-			(get_step(src, direction))
+		// Get step multi-z is cached now so is much faster
+		var/turf/current_turf = get_step_multiz(src, direction)
```
Comment (itsmeow) [inline code/modules/atmospherics/environmental/LINDA_system.dm:54]: It can't be inlined anymore? It was inlined for performance as it actually saves quite a lot. This proc is heavy, cached or not

File: code/game/turfs/change_turf.dm [hunk 5]
```diff
@@ -105,8 +107,24 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 	//We do this here so anything that doesn't want to persist can clear itself
 	var/list/old_comp_lookup = comp_lookup?.Copy()
 	var/list/old_signal_procs = signal_procs?.Copy()
+
+	var/old_atoms_state = SSatoms.initialized
+	SSatoms.initialized = INITIALIZATION_INSSATOMS
```
Comment (itsmeow) [inline code/game/turfs/change_turf.dm:112]: Is this a regression? Why is this here?

**Review by itsmeow — CHANGES_REQUESTED** [review]
Edit zmimic.md to explain that nonlinear stacks are now supported.

File: code/__DEFINES/multi_z.dm (hunk 1 above)
Comment (itsmeow) [reply code/__DEFINES/multi_z.dm]: is this resolved or not important currently?

File: code/game/turfs/change_turf.dm (hunk 5 above)
Comment (PowerfulBacon) [reply code/game/turfs/change_turf.dm:112]: Because if you change turf during mapload, the turf that was changed will never be initialised since it is created but is either never added to the atom initialisation list or may be added twice in some cases due to turf references transfering to the new turf. This will just force the turf to always initialise instantly when calling changeturf. I can't exactly remember quite why it was needed, but I think it was because z-mimic checks for the initialisation state before setting up z-mimic and the atom was being initialised while SSatoms was in an incorrect mode. I still can't remember exactly why it was needed, but I can take a look at it again in code and see what the problem was without it.

Edit:
Looking at the commit text, I believe there was an underlying issue with z-mimic that this fixed but I can't remember quite what it is. Will update if I get any more information.

File: code/modules/atmospherics/environmental/LINDA_system.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/modules/atmospherics/environmental/LINDA_system.dm:54]: It's a lot more challenging since the function holds a local state, however this function doesn't seem to work with non-cardinal directions so if that assumption holds then its probably technically possible

File: code/game/turfs/change_turf.dm (hunk 5 above)
Comment (itsmeow) [reply code/game/turfs/change_turf.dm:112]: I'm worried this will break some niche types of maploading

**Comment by BriggsIDP** [comment]
Some issues I noticed with the PR during the event, but it's possible it was just our implantation of the PR:
- You cannot stack V-Zs and see though more then one layer down, it's just back
image
image
- Parallaxing did not seem to work when looking down at a V-Z.
- Not sure what cased this, but this
image
- Unable to drag mobs up/across V-Zs

**Comment by Admiral-Gigaton** [comment]
how would multi floor z level shuttles even work??

File: code/modules/atmospherics/environmental/LINDA_system.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/modules/atmospherics/environmental/LINDA_system.dm:54]: I have inlined the entire proc, meaning that all usecases will get the advantage of being inlined

File: code/__HELPERS/multiz_helpers.dm [hunk 6]
```diff
@@ -108,8 +108,6 @@ GLOBAL_VAR_TYPED(temporary_multiz_step_ref, /turf)
 	if (above)
 		above.z_depth = null
 		above.below = null
-		// Cleanup z-mimic stuff as it is no longer being used
-		above.cleanup_zmimic()
```
Comment (itsmeow) [inline code/__HELPERS/multiz_helpers.dm]: This would result in the old turf holding a mimic of the previous below turf, would it not?

File: code/__HELPERS/multiz_helpers.dm [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	else if (!above_turf)
+		return null
+	above = above_turf
+	if (!above_turf)
+		calculate_zdepth()
+		return null
+	above_turf.below = src
+	var/was_enabled = above.shadower
+	if (was_enabled)
+		above.cleanup_zmimic()
```
Comment (itsmeow) [inline code/__HELPERS/multiz_helpers.dm:120]: why does this only apply to the 'new' above, shouldn't the old above also cleanup?

File: code/controllers/subsystem/zcopy.dm [hunk 8]
```diff
@@ -200,7 +155,7 @@ SUBSYSTEM_DEF(zcopy)
 				break
 			continue
 
-		// If we're not at our most recent queue position, don't bother -- we're updating again later anyways.
+		// If we're not at our most recent queue position, don't bother -- we're updating again later anyways.aw
```
Comment (itsmeow) [inline code/controllers/subsystem/zcopy.dm]:
```suggestion
		// If we're not at our most recent queue position, don't bother -- we're updating again later anyways.
```

File: code/modules/atmospherics/environmental/LINDA_system.dm [hunk 9]
```diff
@@ -85,18 +76,9 @@
 		return
 	if(disable_adjacent)
 		// I am essentially inlineing two get_dir_multizs here, because they're way too slow on their own. I'm sorry brother
-		var/list/z_traits = SSmapping.multiz_levels[z]
 		for(var/direction in GLOB.cardinals_multiz)
 			// Yes this is a reimplementation of get_step_mutliz. It's faster tho.
```
Comment (itsmeow) [inline code/modules/atmospherics/environmental/LINDA_system.dm:80]:
```suggestion
```

File: code/modules/multiz/zmimic/mimic_turf.dm [hunk 10]
```diff
@@ -49,10 +53,7 @@
 		CRASH("Attempt to enable Z-mimic on already-enabled turf!")
 	shadower = new(src)
 	SSzcopy.openspace_turfs += 1
-	var/turf/under = GET_TURF_BELOW(src)
-	if (under)
-		below = under
-		below.above = src
+	GET_TURF_BELOW(src)
```
Comment (itsmeow) [inline code/modules/multiz/zmimic/mimic_turf.dm:56]: does this do anything??

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
Still need to update zmimic documentation

File: code/modules/multiz/zmimic/mimic_turf.dm (hunk 10 above)
Comment (PowerfulBacon) [reply code/modules/multiz/zmimic/mimic_turf.dm:56]: Get turf below will automatically cache the turf that it finds underneath itself, it just initialises the value in it

File: code/__DEFINES/_globals.dm [hunk 11]
```diff
@@ -53,6 +53,9 @@
 /// Create an untyped null global
 #define GLOBAL_VAR(X) GLOBAL_RAW(/##X); GLOBAL_UNMANAGED(X)
 
+/// Create an untyped null global
+#define GLOBAL_VAR_TYPED(X, Typepath) GLOBAL_RAW(Typepath/##X); GLOBAL_UNMANAGED(X)
```
Comment (PowerfulBacon) [inline code/__DEFINES/_globals.dm:57]: This is just the same as GLOBAL_DATUM, albeit with a better name and an incorrect comment

**Comment by EvilDragonfiend** [comment]
> Things are just randomly disappearing, why!!!

That's because `mimic/var/destruction_timer` is randomly queued even if it shouldn't

* https://github.com/BeeStation/BeeStation-Hornet/pull/11215

I tried to fix it, but now then `mimic/var/depth` is not calculated correctly that causes incorrect `atom/plane` value in the PR.
